### PR TITLE
test: debug transaction leak

### PR DESCRIPTION
**Pull request by SimonSchick**
Dummy PR to trigger CI, please ignore

### CODE REVIEW

**Comment by SimonSchick** [comment]
The single time I want things to fail, they don't....

**Comment by SimonSchick** [comment]
@sushantdhiman I ran this about 20 times now, no failures, any ideas?

**Comment by SimonSchick** [comment]
@sushantdhiman any objections of pulling these changes into master for a bit?
I'm having 0 luck reproducing errors at this time, the 'api' I added should probably not to public hence not docs.

**Comment by jedwards1211** [comment]
@SimonSchick this is related to the CI test failures in my remove bluebird branch right?  I'm kinda mystified by the queries still running errors.

**Comment by SimonSchick** [comment]
@jedwards1211 correct, so far it seems that that some (empty) transaction gets committed outside of a test, but when trying to trace it (see new Error().stack), it works fine.

I'm trying to narrow down what is going on here by enforcing builds to obey package-lock.json and updating dev dependencies (just in case).

Starting to think the way I try to trace things might cause the error not to occur due to timing changes 😞

**Comment by jedwards1211** [comment]
Yeah such are heisenbugs 🙁
I was looking through the code, especially SQLite last night, to see if I would notice anything being unawaited, but so far there hasn't been any smoking gun.

File: lib/sequelize.js [hunk 1]
```diff
@@ -644,11 +644,11 @@ class Sequelize {
       const query = new this.dialect.Query(connection, this, options);
 
       try {
-        await this.runHooks('beforeQuery', options, query);
-        await checkTransaction();
+        await this.runHooks('beforeQuery', options, query, sql);
+        checkTransaction();
```
Comment (eseliger) [inline lib/sequelize.js:648]:
```suggestion
        await checkTransaction();
```

just trying to see if the extra tick in between will cause the error to come back ...

File: test/integration/support.js [hunk 2]
```diff
@@ -2,14 +2,29 @@
...
+      if (queryList) {
+        queryList.push(sql);
+      } else {
+        runningTransactions.set(options.transaction.id, [sql]);
+      }
+    }
+  });
+  this.sequelize.addHook('afterQuery', (options, query, sql) => {
     runningQueries.delete(query);
+    if (options.transaction && sql.includes('COMMIT')) {
```
Comment (eseliger) [inline test/integration/support.js:25]: should this better be:?

```suggestion
    if (options.transaction && ['COMMIT', 'ROLLBACK'].includes(sql)) {
```

File: lib/sequelize.js [hunk 3]
```diff
@@ -644,11 +644,11 @@ class Sequelize {
       const query = new this.dialect.Query(connection, this, options);
 
       try {
-        await this.runHooks('beforeQuery', options, query);
+        await this.runHooks('beforeQuery', options, query, sql);
         await checkTransaction();
```
Comment (eseliger) [inline lib/sequelize.js:648]:
```suggestion
       checkTransaction();
```

**Comment by jedwards1211** [comment]
I was looking at `retry-as-promised`:

```js
    Promise.resolve(callback({ current: options.$current }))
      .then(resolve)
      .then(function() {
        if (timeout) clearTimeout(timeout);
        if (backoffTimeout) clearTimeout(backoffTimeout);
      })
      .catch(function(err) {
        if (timeout) clearTimeout(timeout);
        if (backoffTimeout) clearTimeout(backoffTimeout);
```
This is probably a dumb idea but I don't guess there's a way the timeout could fire again between `.then(resolve)` and what comes after it?

**Comment by SimonSchick** [comment]
@jedwards1211 I'm unsure, might be worthwhile to introduce a test case directly against that library.
I will merge this now (and revert it later) to see if we can repro this mess in other branches.

**Comment by sushantdhiman** [comment]
Why was this merged, it just pollute commit history. If something can't be replicated in this branch how merging in master would replicate that issue

**Comment by jedwards1211** [comment]
@SimonSchick mentioned wanting to try to repro in other branches, so I guess the idea is to pull this into those branches.  I think there are ways to pull in changes from this branch since master into another branch, but I forget how to do it...

**Comment by sushantdhiman** [comment]
If this branch can't replicate that failure even once how it will help in other branches

**Comment by SimonSchick** [comment]
Me and @eseliger were both pretty baffled why we couldn't replicate the error with those changes as functionality nothing was changed.

We came up with the idea to have this check run on more builds and hopefully come across one where this condition occurs rather than mindlessly hammering travis over and over and wasting a LOT of time in the process.

It's not 100% not ideal but doesn't really affect anything outside of CI.

**Comment by sushantdhiman** [comment]
@SimonSchick This PR worked by the way. It seems https://github.com/sequelize/sequelize/blob/master/test/integration/cls.test.js#L67 is culprit

**Comment by SimonSchick** [comment]
Cool, will file a PR.

**Pull request merged**